server/src/main/java/org/elasticsearch/search/aggregations/metrics/WrapperTDigestArrays.java

```diff
@@ -0,0 +1,365 @@
+                breaker.addWithoutBreaking(delta);
+                throw e;
+            }
+        } else {
+            breaker.addWithoutBreaking(delta);
+        }
+    }
+
+    private abstract static class MemoryAwareArray implements Releasable, Accountable {
+        protected final CircuitBreaker breaker;
+        private final AtomicBoolean closed = new AtomicBoolean(false);
+
+        MemoryAwareArray(CircuitBreaker breaker) {
+            this.breaker = breaker;
+        }
+
+        @Override
+        public final void close() {
+            if (closed.compareAndSet(false, true)) {
+                breaker.addWithoutBreaking(-ramBytesUsed());
+            }
+        }
+    }
+
+    public static class WrapperTDigestDoubleArray extends MemoryAwareArray implements TDigestDoubleArray {
+        static final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(WrapperTDigestDoubleArray.class);
+
+        private double[] array;
+        private int size;
+
+        public WrapperTDigestDoubleArray(CircuitBreaker breaker, int initialSize) {
+            this(breaker, new double[initialSize]);
+        }
+
+        public WrapperTDigestDoubleArray(CircuitBreaker breaker, double[] array) {
+            super(breaker);
+            this.array = array;
+            this.size = array.length;
+        }
+
+        @Override
+        public long ramBytesUsed() {
+            return SHALLOW_SIZE + RamUsageEstimator.sizeOf(array);
+        }
+
+        @Override
+        public int size() {
+            return size;
+        }
+
+        @Override
+        public double get(int index) {
+            assert index >= 0 && index < size;
+            return array[index];
+        }
+
+        @Override
+        public void set(int index, double value) {
+            assert index >= 0 && index < size;
+            array[index] = value;
+        }
+
+        @Override
+        public void add(double value) {
+            ensureCapacity(size + 1);
+            array[size++] = value;
+        }
+
+        @Override
+        public void sort() {
+            Arrays.sort(array, 0, size);
+        }
+
+        @Override
+        public void resize(int newSize) {
+            ensureCapacity(newSize);
+
+            if (newSize > size) {
+                Arrays.fill(array, size, newSize, 0);
+            }
+
+            size = newSize;
+        }
+
+        @Override
+        public void ensureCapacity(int requiredCapacity) {
+            if (requiredCapacity > array.length) {
+                int newSize = array.length + (array.length >> 1);
+                if (newSize < requiredCapacity) {
+                    newSize = requiredCapacity;
+                }
+                double[] newArray = new double[newSize];
+                System.arraycopy(array, 0, newArray, 0, size);
+
+                long oldRamBytesUsed = ramBytesUsed();
+                array = newArray;
+                breaker.addWithoutBreaking(ramBytesUsed() - oldRamBytesUsed);
```

Comment:
Ditto, move this before allocating `newArray` so that the breaker trips before an allocation that may lead to OOM.

Reply:
Done!